src/zarr/core/indexing.py

```diff
@@ -1378,7 +1386,17 @@ def is_total_slice(item: Selection, shape: ChunkCoords) -> bool:
                 isinstance(dim_sel, slice)
                 and (
                     (dim_sel == slice(None))
-                    or ((dim_sel.stop - dim_sel.start == dim_len) and (dim_sel.step in [1, None]))
+                    or (
+                        dim_sel.stop is not None
+                        and (dim_sel.stop - dim_sel.start == dim_len)
+                        and (dim_sel.step in [1, None])
```

Comment:
it's kinda of weird that the step matters at all, if we know the stop. I understand this is how slice objects work, but maybe zarr indexing should not rely on python slice object semantics.

Reply:
but you can do `a[slice(None, None, 2] = 1` for example...

Reply:
`__getitem__` should totally accept slices as input, but we should _immediately_ convert slices into objects that make sense for zarr array indices (finite sets of regularly spaced non-negative integers). Then we don't have to handle `Nones` for checks like this here, and we could frame the problem in terms of sets of integers. Like, "is_total_slice" is equivalent to "is the set of indices for the chunk completely contained by this other set of indices"

Reply:
(out of scope for this PR of course)

Reply:
> "is_total_slice" is equivalent to "is the set of indices for the chunk completely contained by this other set of indices"

right but then handling the potentially-smaller-last-chunk is painful. Today's logic is basically that, except it can't handle a smaller last chunk because it has no context.